### ISSUE TITLE
issue retraining the model with a small data set for specific classes ?

### ISSUE THREAD

**Issue by vipin-prabhakaran** [ultralytics/ultralytics#9037]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have a YOLOV8 model trained with 8000 images and 15 classes .while testing we observed some classes are not detecting properly and some false detections ,we are using ultralytics HUB for training .there no option for retraining the existing model is there .so I have created a new data set from Roboflow  with 4 classes and added 100 images and labels .now I want to retrain my existing model for this classes with out losing the existing classes and its indexes.
I have retrained the model with following command 
yolo task=detect mode=train model=/Users/Desktop/retrain18/v17ppe.pt imgsz=640 data=/Users/Desktop/retrain18/data.yaml epochs=100

after that I have obtained tflite model ,but its giving wrong class names on detection .

can any one suggests  where I am going wrong on these steps ,or which step I am missing for properly retraining the existing model?

**Comment by Burhan-Q** [comment]
@vipin-prabhakaran There is no process for "additive" training of a model. If you change the number of classes or want to train on a dataset with more samples, you will need to start a new training session. 

This is the same for the case when you collect 1000 images (dataset A), train a model, then collect another 1000 images (dataset B) and attempt to train on only the new images (dataset B), this will not "add" to the result from the prior training. The only information retained from the last training is whatever was gained from the final training, the weights from the first training (dataset A) will be overwritten.

**Comment by vipin-prabhakaran** [comment]
Thanks for your response .
can you pls help me to find out how to retrain /fine tune my model with updated images for some classes ,I have created a merged data set which contains both old & new data.now my issue is how to retrain /fine tune the model using CLI command to keep the existing trained data intact + retrained for the classes where I added new images .basically I modified images for the existing classes only .

**Comment by Burhan-Q** [comment]
@vipin-prabhakaran you may have misunderstood my prior message. I'll try to be clearer by saying your statement:

> keep the existing trained data intact + retrained for the classes where I added new images

is **not possible.** When you have the combined dataset, you need to start training with the new dataset as **new**. Any attempt to "retrain" or "add training" will erase what the model learned previously.

**Comment by vipin-prabhakaran** [comment]
@Burhan-Q  thanks for your response ,but I am really now concerned about how can I keep my model go for continuous training or fine tuning .as per your suggestion I need to always train my model from scratch whenever I am adding /deleting images for any class ,can you please clear my confusion in this

**Comment by Burhan-Q** [comment]
@vipin-prabhakaran you will have to retrain the model on the entire dataset you wish to have the model trained on. The model has no "memory" and when training your model it will only train on the data that is present. If you train 5,000 images, then add 3,000 images, you **must** train on ALL 8,000 images.

**Comment by vipin-prabhakaran** [comment]
@Burhan-Q  so as per your reply my assumption is I had a model which is trained for 100 epocs on a data set with 8000 images ,
now I added 200 images to the same data set ,so I will retrain my model with a data set of 8200 images for 10 epocs with the command   
" yolo task=detect mode=train data=/Users/Desktop/retrain18/data.yaml epochs=10 imgsz=640 model=/Users/Desktop/retrain18/v17ppe.pt plots=True "

so my conclusion is now my model is totally had the data of 110 epocs (100 epoc with old data set + 10 epoc with new data set ),after this my model accuracy will be more accurate ,pls correct me if any of conclusion is wrong

**Comment by Burhan-Q** [comment]
@vipin-prabhakaran this is incorrect 

> so my conclusion is now my model is totally had the data of 110 epocs (100 epoc with old data set + 10 epoc with new data set)

You would _only_ have a model trained on 8,200 images for 10 epochs. If you want to have a model trained for 110 epochs, you must train it for 110 epochs `epochs=110` once the model is finished training, it is **finished** and it's not possible to "add" anything to that model.

**Comment by vipin-prabhakaran** [comment]
So as per this reply  there is no way I can start my training by keeping my already trained model as base model and keep on improving and fine tuning my trained model .its always a fresh training with my updated data set for which ever cycle I am giving ,it will be of great help  if you can explain me what's the idea of retraining /finetuning  of yolo model as I am newbie in this yolo training

**Comment by glenn-jocher** [comment]
@vipin-prabhakaran hello! 😊

In YOLO and similar deep learning models, "retraining" or "finetuning" essentially means starting the new training session with the weights of a previously trained model, rather than from scratch. It doesn't concatenate training periods (epochs) but leverages learned patterns to potentially speed up learning on new or expanded data.

When you add new images or modify your existing dataset, you'll initiate training with these commands, specifying your previously trained model (`v17ppe.pt` in your case) as the starting point:

```bash
yolo task=detect mode=train data=/Users/Desktop/retrain18/data.yaml epochs=10 imgsz=640 model=/Users/Desktop/retrain18/v17ppe.pt
```

This uses the learned weights from `v17ppe.pt` and attempts to refine (finetune) them based on your new/combined dataset. Just remember, the training process will adapt to the entire dataset provided, enhancing the model's understanding of both old and newly added data.

Happy training! 🚀

**Comment by jbdebelle** [comment]
@Burhan-Q, I'd appreciate some clarification because what you're saying appears to conflict with the message that @glenn-jocher has posted below and what has been discussed in https://github.com/ultralytics/ultralytics/issues/9944.

I'm inclined to trust your information more, especially since much of what @glenn-jocher says seems to be generated by GPT.

If I've fine-tuned a model on a dataset with two classes using the following command:

````
yolo task=detect mode=train data=path_data1/data.yaml epochs=100 imgsz=640 model=yolov8.pt pretrained=true
````
and saved the output in a model named saved_finetuned_model.pt, and then I attempt to fine-tune these weights on another dataset containing new images with the same two classes using this command:

````
yolo task=detect mode=train data=path_data2/data.yaml epochs=110 imgsz=640 model=saved_finetuned_model.pt pretrained=true
````

Are you saying that the second fine-tuning process would essentially just train the yolov8n.pt model for 110 epochs on path_data2/data.yaml, effectively forgetting everything it had learned previously?

Thank you for your clarification.

**Comment by Burhan-Q** [comment]
@jbdebelle you're correct in your assessment that the second training will cause the model to "forget" everything from the first training. This is because every parameter of the model will be updated based on the information of the dataset provided at training time. 

What you _could_ do instead is freeze all but the final layer on the **second** training, and train the model on all the data with both classes. This would work if the first model was already successful at detecting objects and just need to train to improve the classification. 

```bash
yolo task=detect mode=train data=path_data2/data.yaml freeze=22 epochs=110 imgsz=640 model=saved_finetuned_model.pt pretrained=true
```

Freezing would avoid any "forgetting" in the detection layers of the model parameters, and would also let the second training run much quicker. You want to keep all the samples in the training because the model will need to "finetune" on top of the previous weights; really _I think_ this should be referred to as `transfer-learning` but LLM terminology has bled into the CV space. Note that without the previous samples, the weights could drastically change if the new data is different enough (you could test this, but you'd likely end up doing another training anyhow and wouldn't save any time/efforts).

**Comment by jbdebelle** [comment]
@Burhan-Q.

 I have a dataset of people taken with one camera, and for the second dataset, it's a similar environment with images of people, but the camera optics are slightly different. I want my model to adapt to these new optics, but the content of the images is very similar.

I will try freezing the layers as you have recommended. 
Out of curiosity, if I use a small learning rate and don't freeze the layers, given that the data is quite similar in both cases, will the model really "forget" everything from the first training, or is this more of a figure of speech meant for cases where the data is very different?

Thanks for the answer.
